plugins/serialization/src/main/java/org/qbicc/plugin/serialization/BuildtimeHeap.java

```diff
@@ -35,39 +47,187 @@ public static BuildtimeHeap get(CompilationContext ctxt) {
+                            memberMap.put(member, lf.literalOf(jf.getDouble(instance)));
+                        } else {
+                            Object fieldContents = jf.get(instance);
+                            if (fieldContents == null) {
+                                memberMap.put(member, lf.zeroInitializerLiteralOfType(member.getType()));
+                            } else {
+                                SymbolLiteral childSymbol = serializeObject(fieldContents);
+                                memberMap.put(member, lf.bitcastLiteral(childSymbol, (WordType)member.getType()));
+                            }
+                        }
+                    } catch (IllegalAccessException e) {
+                        ctxt.error("Heap Serialization: denied access to field %s of %s", jf.getName(), jClass);
+                    }
+                }
+            }
+            qClass = qClass.getSuperClass();
+            jClass = jClass.getSuperclass();
         }
+
+        Literal objLiteral = ctxt.getLiteralFactory().literalOf(objType, memberMap);
+        SymbolLiteral objName = lf.literalOfSymbol(myName, objType.getPointer().asCollected());
+        initialHeap.put(objName, objLiteral);
+
+        return objName;
     }
 
-    public byte[] getHeapBytes() {
-        return ser.getBytes();
+    private SymbolLiteral serializeArray(byte[] array) {
+        TypeSystem ts = ctxt.getTypeSystem();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+        String myName = nextLiteralName();
+
+        LoadedTypeDefinition arrayLTD = layout.getByteArrayContentField().getEnclosingType().load();
+        Layout.LayoutInfo objLayout = layout.getInstanceLayoutInfo(arrayLTD);
+        CompoundType arrayCT = objLayout.getCompoundType();
+
+        CompoundType.Member contentMem = objLayout.getMember(layout.getByteArrayContentField());
+        ArrayType s8ArrayType = ts.getArrayType(ts.getSignedInteger8Type(), array.length);
+        CompoundType.Member realContentMem = ts.getCompoundTypeMember(contentMem.getName(), s8ArrayType, contentMem.getOffset(), contentMem.getAlign());
+
+        CompoundType literalCT = ts.getCompoundType(CompoundType.Tag.STRUCT, myName+"_type",arrayCT.getSize() + s8ArrayType.getSize(),
+            arrayCT.getAlign(), () -> List.of(arrayCT.getMember(0), arrayCT.getMember(1), realContentMem));
+
+        Literal arrayLiteral = lf.literalOf(literalCT, Map.of(
+            literalCT.getMember(0), lf.literalOf(arrayLTD.getTypeId()),
+            literalCT.getMember(1), lf.literalOf(array.length),
+            realContentMem, lf.literalOf(s8ArrayType, array)
+        ));
+        SymbolLiteral arrayName = lf.literalOfSymbol(myName, literalCT.getPointer().asCollected());
+        initialHeap.put(arrayName, arrayLiteral);
+
+        return arrayName;
     }
 
-    public List<InitializedField> getHeapRoots() {
-        return heapRoots;
+    private SymbolLiteral serializeArray(boolean[] array) {
```

Comment:
Leaving the other 8 array cases stubbed out until we get the byte[] case which is used by String to actually do what we want.